torchtune/utils/perf_utils.py

```diff
@@ -0,0 +1,117 @@
+class _MetricTacker:
+    count: int = 0
+    val: float = 0.0
+
+
+def _compute_rolling_average(prev_avg, val, n) -> float:
+    return prev_avg * (n - 1) / n + val / n
+
+
+# Only support for one reduction type. And the same metric can't have multiple reduction.
+
+
+class TunePerfMonitor:
+    def __init__(self):
+        self._metric_dict: Dict[str, _MetricTacker] = defaultdict(_MetricTacker)
+        self._inflight: Dict[str, float] = {}
+
+    def _run_rolling_average_reduction(self, metric_name: str, metric_val: float) -> float:
+        # Rolling average
+        metric_info = self._metric_dict[metric_name]
+        count, prev_avg = metric_info.count + 1, metric_info.val
+        avg = _compute_rolling_average(prev_avg=prev_avg, val=metric_val, n=count)
+        return avg
+
+    def _run_reduction(
+        metric_name: str, metric_val: float, reduction_type: _ReductionType
+    ) -> float:
+        # TODO: if/else on reduction_type is not scalable. Update to using a dispatch
+        # when we have more reduction types.
+        if self.reduction_type == _ReductionType.MEAN:
+            if metric_name not in self._metric_dict:
+                return metric_val
+            else:
+                return self._run_rolling_average_reduction(
+                    metric_name=metric_name, metric_val=metric_val
+                )
+
+        raise RuntimeError(f"Reduction type {reduction_type} not supported!")
+
+
+    def _record_metric(slef, metric_name, metric_val):
+        """
+        Updates our metric tracking for metric_name with metric_val and bumps count by 1.
+        """
+        if metric_name not in self._metric_dict:
+            self._metric_dict[metric_name] = _MetricTacker(count=1, val=metric_val)
+        else:
+            self._metric_dict[metric_name].count += 1
+            self._metric_dict[metric_name].val = metric_val
+
+    def start_record(metric: str):
+        """
+        Start tracking metric given by ``str``
+        """
+        assert (
+            metric not in self._inflight
+        ), f"Already tracking metric {metric} for this iteration!"
+        self._inflight[metric] = time.perf_counter()
+
+    def end_record(
+        metric_name: str, reduction_type: _ReductionType = _ReductionType.MEAN
+    ):
+        """
+        End tracking metric given by ``str`` and compute and store the metric value.
+        """
+        # TODO: only MEAN reduction supported right now, and end_record is not the best place to specify it.
+        # Don't want to specify it in emit_metric either, as then we'll have to store metric for every
+        # iteration, possibly too much space used. Could have a solution where user a priori defines the metric
+        # they want to track and the reduction type for it.
+        assert (
+            metric_name in self._inflight
+        ), f"Not tracking metric {metric_name} for this iteration!"
+        cpu_elapsed_s = time.perf_counter() - self._inflight.pop(metric_name)
```

Comment:
I think the start_record, end_record design here only supports  iteration time related logging? How it fit into pytorch profiler, memory snapshot etc as well?

Reply:
@SLR722 , `log_metric` is intended to support the use case of emitting a metric that's only computed with a single data point / at a single point in time. 

> fit into pytorch profiler, memory snapsho

So IMO `TunePerfMonitor` is a way to track and retrieve metrics. Profiler and snapshot are related but seperate tools. I currentl don't see the value in integrating all of these into a single class, though I'm open to feedback and understand the motivation for that design. Could also be open to a solution of designing something like a `snapshot_memory` API within this perf monitor to take the memory snapshot, but I do feel like that it takes away from the simplicity of this class.